### PR TITLE
AddRunParameters files to Housekeeper when finishing flow cell

### PR DESCRIPTION
**Pull request by diitaz93**
## Description
Makes the command `finish flow-cell` store the run parameters file in Housekeeper.

### Added

- Tag `run-parameters`
- Function `add_run_parameters_file_to_housekeeper` in `cg/meta/demultiplex/housekeeper_storage_functions.py`
- Test for `add_run_parameters_file_to_housekeeper`
- A check in the test of `finish flow-cell` that verifies that the run parameters file has been added to Housekeeper.

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b add-run-params-hk -a
    ```

### How to test

- [x] Do `cg -l DEBUG demultiplex finish flow-cell 20231108_LH00188_0028_B22F52TLT3`
- [x] Verify that the run parameters has been added to Housekeeper
```shell
Created new bundle version dir: /home/proj/stage/housekeeper-bundles/22F52TLT3/2023-11-08
Linked file: /home/proj/stage/sequencing_data/illumina/flow_cells/20231108_LH00188_0028_B22F52TLT3/RunParameters.xml -> /home/proj/stage/housekeeper-bundles/22F52TLT3/2023-11-08/RunParameters.xml
File added to Housekeeper bundle 22F52TLT3
Added run parameters file /home/proj/stage/sequencing_data/illumina/flow_cells/20231108_LH00188_0028_B22F52TLT3/RunParameters.xml to 22F52TLT3 in Housekeeper.
[js.diazboada@hasta:demultiplexed-runs] [S_base] $ housekeeper get bundle -v 22F52TLT3
2024-04-29 10:10:06 hasta.scilifelab.se housekeeper.cli.core[209368] INFO Use root path /home/proj/stage/housekeeper-bundles
                📦 Bundle table 📦                 
┏━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ ID     ┃ Bundle name ┃ Version IDs ┃ Created    ┃
┡━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
│ 159280 │ 22F52TLT3   │ 163391      │ 2023-11-08 │
└────────┴─────────────┴─────────────┴────────────┘
                              📕 Version table 📕                               
┏━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━┓
┃ ID     ┃ Bundle name ┃ Nr files ┃ Included ┃ Archived ┃ Created    ┃ Expires ┃
┡━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━┩
│ 163391 │ 22F52TLT3   │ 4        │          │          │ 2023-11-08 │         │
└────────┴─────────────┴──────────┴──────────┴──────────┴────────────┴─────────┘
                                                       📜 Files table 📜                                                        
┏━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ ID      ┃ File name                                                                              ┃ Tags                      ┃
┡━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ 4870278 │ /home/proj/stage/housekeeper-bundles/22F52TLT3/2023-11-08/SampleSheet.csv              │ samplesheet, 22F52TLT3    │
│ 4878111 │ /home/proj/stage/housekeeper-bundles/22F52TLT3/2023-11-08/22F52TLT3_demultiplex.stdout │ log, 22F52TLT3            │
│ 4878112 │ /home/proj/stage/housekeeper-bundles/22F52TLT3/2023-11-08/22F52TLT3_demultiplex.stderr │ log, 22F52TLT3            │
│ 5882700 │ /home/proj/stage/housekeeper-bundles/22F52TLT3/2023-11-08/RunParameters.xml            │ 22F52TLT3, run-parameters │
└─────────┴────────────────────────────────────────────────────────────────────────────────────────┴───────────────────────────┘
[js.diazboada@hasta:demultiplexed-runs] [S_base] $
```

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [x] **MINOR** - when you add functionality in a backwards compatible manner

## Implementation Plan

- [ ] Document in [Atlas - Demultiplexing](https://atlas.scilifelab.se/production/data_analysis/supporting_procedures/demultiplexing_in_production/demultiplexing/#manual-demultiplexing-and-fastq-generation)
- [ ] Deployed to stage:
  ```shell

  ```
- [ ] Deployed to production:
  ```shell

  ```
- [ ] Inform to @Clinical-Genomics/prodbioinfo

### CODE REVIEW

File: cg/meta/demultiplex/housekeeper_storage_functions.py [hunk 1]
```diff
@@ -92,6 +95,20 @@ def add_demux_logs_to_housekeeper(
...
+) -> None:
+    """Add run parameters file to Housekeeper."""
+    flow_cell_path = Path(flow_cell_run_dir, flow_cell_name)
+    flow_cell = FlowCellDirectoryData(flow_cell_path)
+    run_parameters_file_path: Path = flow_cell.run_parameters_path
+    tag_names: list[str] = [SequencingFileTag.RUN_PARAMETERS, flow_cell.id]
+    hk_api.add_file_to_bundle_if_non_existent(
+        file_path=run_parameters_file_path, bundle_name=flow_cell.id, tag_names=tag_names
+    )
+    LOG.info(f"Added run parameters file {run_parameters_file_path} to Housekeeper.")
```
Comment (islean) [inline cg/meta/demultiplex/housekeeper_storage_functions.py:109]: Maybe a bit over the top with other logs but would it make sense to add the bundle here? For example
```suggestion
    LOG.info(f"Added run parameters file {run_parameters_file_path} to {flow_cell.id} in Housekeeper.")
```

File: tests/meta/demultiplex/test_housekeeper_storage_functions.py [hunk 2]
```diff
@@ -211,6 +212,33 @@ def test_add_demux_logs_to_housekeeper(
         assert file.path.split("/")[-1] in expected_file_names
 
 
+def test_add_run_parameters_to_housekeeper(
+    demultiplex_context: CGConfig, novaseq_x_flow_cell: FlowCellDirectoryData
+):
+    """Test that the run parameters file of a flow cell is added to Housekeeper."""
+    # GIVEN a flow cell with a run parameters file and a housekeeper API
```
Comment (henrikstranneheim) [inline tests/meta/demultiplex/test_housekeeper_storage_functions.py:219]:
```suggestion
    # GIVEN a flow cell with a run parameters file and a Housekeeper API
```

**Review by islean — APPROVED** [review]
Looks good to me!

File: tests/meta/demultiplex/test_housekeeper_storage_functions.py [hunk 3]
```diff
@@ -211,6 +212,33 @@ def test_add_demux_logs_to_housekeeper(
...
+    hk_api = demultiplex_context.housekeeper_api
+
+    # GIVEN that a run parameters file does not exist for the flow cell in housekeeper
```
Comment (henrikstranneheim) [inline tests/meta/demultiplex/test_housekeeper_storage_functions.py:222]:
```suggestion
    # GIVEN that a run parameters file does not exist for the flow cell in Housekeeper
```

File: tests/meta/demultiplex/test_housekeeper_storage_functions.py [hunk 4]
```diff
@@ -211,6 +212,33 @@ def test_add_demux_logs_to_housekeeper(
...
+        flow_cell_name=novaseq_x_flow_cell.full_name,
+        flow_cell_run_dir=demultiplex_context.demultiplex_api.flow_cells_dir,
+        hk_api=hk_api,
+    )
+
+    # THEN the run parameters file was added to housekeeper
+    run_parameters_file: File = hk_api.files(
+        tags=[SequencingFileTag.RUN_PARAMETERS, novaseq_x_flow_cell.id]
+    ).first()
+    assert run_parameters_file.path.endswith("RunParameters.xml")
```
Comment (henrikstranneheim) [inline tests/meta/demultiplex/test_housekeeper_storage_functions.py:239]: Use FileExtension constant.

**Review by henrikstranneheim — APPROVED** [review]
Great, 👍 .

File: tests/meta/demultiplex/test_housekeeper_storage_functions.py (hunk 4 above)
Comment (diitaz93) [reply tests/meta/demultiplex/test_housekeeper_storage_functions.py:239]: fixed!

File: cg/meta/demultiplex/housekeeper_storage_functions.py (hunk 1 above)
Comment (diitaz93) [reply cg/meta/demultiplex/housekeeper_storage_functions.py:109]: I will add it and see how it looks!

Comment (diitaz93) [reply cg/meta/demultiplex/housekeeper_storage_functions.py]: looks good, I will leave it

**Pull request merged**